File: src/tools/shared/imageValidation.ts
import type { ToolContext } from "src/Tool";
import {
  type ImageValidationResult,
  createImageValidator,
  formatImageValidationError,
  imageValidationWarnings,
} from "src/services/docker/imageValidator";

export interface ToolImageValidationResult {
  results: ImageValidationResult[];
  error: string | null;
  warnings: string[];
}

function shouldBlockUnknownImages(): boolean {
  return process.env.DOCKER_AGENT_IMAGE_VALIDATION_UNKNOWN === "block";
}

export async function validateImagesForTool(
  images: string[],
  ctx: ToolContext,
): Promise<ToolImageValidationResult> {
  const validator = ctx.imageValidator ?? createImageValidator(ctx.dockerEngine);
  const uniqueImages = [...new Set(images)];
  const results = await validator.validateImages(uniqueImages, { signal: ctx.abortSignal });
  return {
    results,
    error: formatImageValidationError(results, { blockUnknown: shouldBlockUnknownImages() }),
    warnings: imageValidationWarnings(results),
  };
}

File: servers/docker-mcp-server/src/docker_mcp_server/tools/shared/image_validation.py
"""Tool-level image validation wrapper.
"""

from __future__ import annotations

import os
from dataclasses import dataclass

from docker_mcp_server.services.docker.image_validator import (
    ImageValidationResult,
    create_image_validator,
    format_image_validation_error,
    image_validation_warnings,
)
from docker_mcp_server.tools.base import ToolContext


@dataclass
class ToolImageValidationResult:
    results: list[ImageValidationResult]
    error: str | None
    warnings: list[str]


def _should_block_unknown_images() -> bool:
    return os.environ.get("DOCKER_AGENT_IMAGE_VALIDATION_UNKNOWN") == "block"


async def validate_images_for_tool(
    images: list[str], ctx: ToolContext
) -> ToolImageValidationResult:
    """Validate a deduplicated list of images for tool use."""
    validator = (
        ctx.image_validator
        if ctx.image_validator is not None
        else create_image_validator(ctx.docker_engine)
    )
    unique_images = list(dict.fromkeys(images))
    results = await validator.validate_images(
        unique_images, signal=ctx.abort_signal
    )
    return ToolImageValidationResult(
        results=results,
        error=format_image_validation_error(
            results, block_unknown=_should_block_unknown_images()
        ),
        warnings=image_validation_warnings(results),
    )


__all__ = ["ToolImageValidationResult", "validate_images_for_tool"]
